### DISCUSSION TITLE
Unsubscribing a non-owned subscription

### DISCUSSION THREAD

**Discussion by ecorm** [wamp-proto/wamp-proto#496] [Q&A] [answered]
What is the expected behavior of a Broker when a client sends an UNSUBSCRIBE with the ID of an existing subscription, but for which it never subscribed?

For example

Client 1
```
TX: [SUBSCRIBE,1,{},"topic"]
RX: [SUBSCRIBED,1,42,"topic"] // Subscription ID 42
```
Client 2
```
TX: [UNSUBSCRIBE,1,42] // Subscription ID 42
RX: ?
```

My guess is that the Broker should respond with an ERROR with URI=`wamp.error.no_such_subscription`. While the subscription does indeed exist, the client issuing the UNSUBSCRIBE is not a subscriber to the topic, so it's like saying "there's no such subscription _for you_".

This behavior is also what I would expect from a client unsubscribing twice to a subscription it actually owned:
```
TX: [SUBSCRIBE,1,{},"topic"]
RX: [SUBSCRIBED,1,42,"topic"] // Subscription ID 42
TX: [UNSUBSCRIBE,2,42] // Subscription ID 42
RX: [UNSUBSCRIBED,2]
TX: [UNSUBSCRIBE,3,42] // Subscription ID 42
RX: [ERROR,UNSUBSCRIBE,3,{},"wamp.error.no_such_subscription"]
```

This way, the broker always applies the same logic without having to remember that a client previously had a subscription to the topic it wants to unsubscribe from.

The same question applies to unregistering an non-owned registration.

Please let me know if my reasoning is correct.

**Comment by Jopie64** [comment]
I think an UNSUBSCRIBE or UNREGISTER should _never_ have any effect on any client other than itself. DEALER should be implemented such that client's can't know about each other's subscriptions/registrations. Registration/subscription ID's of different clients should even be able to overlap.

Also my first hunch about sending an UNSUBSCRIBE to a subscription it doesn't currently own should result in a protocol violation? There can't be races here because of network since a client initiates subscribe/unsubscribe (or registrations) itself. Unless the meta api was used to forcefully unsubscribe someone. So maybe only in that case, no protocol violation since now there can be races?

**Reply by ecorm** [reply, marked as answer]
> DEALER should be implemented such that client's can't know about each other's subscriptions/registrations.

The way I understand it, there is only one subscription ID per topic+match_policy combo, as confirmed by the [`wamp.subscription.lookup`](https://wamp-proto.org/wamp_latest_ietf.html#name-wampsubscriptionlookup) meta procedure.

Several subscribers can associated with the same subscription ID, as confirmed by the [`wamp.subscription.list_subscribers`](https://wamp-proto.org/wamp_latest_ietf.html#name-wampsubscriptionlist_subscr) meta procedure which takes a `subscription|id` argument and returns a list of subscriber session IDs.

So subscriber IDs are not unique to each subscriber as far as the meta API is concerned.

> Unless the meta api was used to forcefully unsubscribe someone. So maybe only in that case, no protocol violation since now there can be races?

So returning an `ERROR` (instead of a protocol violation abort) makes WAMP more future-proof if such a meta procedure is introduced. A `wamp.subscription.remove_subscriber` meta-procedure has been proposed in #253.

**Reply by Jopie64** [reply]
Hmmm... That indeed looks like the model the meta model is exposing indeed... It feels weird to me though...

The model I had in mind was different. If you put it in SQL table terms it would look more or less like this:

```mermaid
erDiagram
  TOPIC {
     int id
     string uri
  }
  SESSION {
    int id
    string ipAddress
    other properties
  }
  SUBSCRIPTION {
     int id
     int idSession
     int idTopic
  }
  TOPIC ||--|{ SUBSCRIPTION: has
  SESSION ||--o{ SUBSCRIPTION: has
```

This way a sessions subscription to a topic is a totally separate object than another session's subscription to the same topic, of which each hold their own id.
But in the other model, it seems the topic and subscription 'table' seem merged together, which feels odd to me...

In the latter case, why have a subscription id (which is then also a topic id) communicated at all? Why not just always reference topics by Uri's, even when doing unsubscribe?

**Reply by ecorm** [reply]
> In the latter case, why have a subscription id (which is then also a topic id) communicated at all? Why not just always reference topics by Uri's, even when doing unsubscribe?

The topic URI is not enough to unsubscribe. A subscription is associated with a topic URI + match policy combination, where match policy is `exact`, `prefix`, or `wildcard`.

**Reply by ecorm** [reply]
> There can't be races here because of network since a client initiates subscribe/unsubscribe (or registrations) itself. Unless the meta api was used to forcefully unsubscribe someone. So maybe only in that case, no protocol violation since now there can be races?

Another future possibility is if router-side session resuming is ever specified, and the client wants to unsubscribe after resuming when a previous unsubscribe attempt was interrupted by a network failure. So again, returning an ERROR instead of treating a non-owned unsubscription ID as a protocol violation makes things more future-proof.

**Reply by oberstet** [reply]
> The way I understand it, there is only one subscription ID per topic+match_policy combo

yes, correct

> Why not just always reference topics by Uri's, even when doing unsubscribe?

because as @ecorm notes, that isn't enough - because of pattern based subs/regs. also, the possible future use he mentions makes a lot of sense.

**Reply by Jopie64** [reply]
> Another future possibility is if router-side session resuming is ever specified, and the client wants to unsubscribe after resuming when a previous unsubscribe attempt was interrupted by a network failure. So again, returning an ERROR instead of treating a non-owned unsubscription ID as a protocol violation makes things more future-proof.

Yea I'm convinced now that indeed it should _not_ result in a protocol violation, mostly because of this future use case.

That `TOPIC` table was wrong indeed. It should probably have been something like a `SubscriptionPattern` or something.

For the rest I have to rethink what the implications are of the, in my view, other model.
At least one I can think of is that a session cannot subscribe twice on the same euhm... how to call it? subscription pattern (😜). Since it would result in the same subscription id.

That one is of course not a big deal because, why would you want to do that? Still, it's an implication nonetheless and not an obvious one...

**Reply by oberstet** [reply]
> a session cannot subscribe twice on the same

yes, true. and in fact, a router needs to handle such cases properly, pls see `was_already_subscribed` here:

https://github.com/crossbario/crossbar/blob/03d7057ee18a0d1e59b33bf1633ee246c1cf908a/crossbar/router/broker.py#L833

https://github.com/crossbario/crossbar/blob/master/crossbar/router/observation.py#L221

in general I think a spec section discussing such consequences and details might indeed be worth.

maybe even a bit of history ... why we arrived at the specific design of this stuff: WAMP tries to hit a sweet spot, freeing app developers from language and network boundaries, while at the same time allowing routers _in the real world_ to provide efficient and fast implementations.

so we do have prefix and wildcard patterns, but we do not have reg ex patterns for example. the former two can be done with prefix indexes / tries efficiently, the latter not (in general).

and having a sub/reg ID saves bytes on the wire, and allow efficient routing (a router needs an object for a reg/sub independent of single sessions), etc etc

**Comment by KSDaemon** [comment]
My thoughts on topic:

1. I like @ecorm's proposal and have the same in my mind.
2. Don't forget that there are IDs of different scopes, and Subscriptions are of router scope and not a session scope. So the same subscription may be returned to different clients.
3. In implementations that I'm aware of - it is impossible to unsubscribe from a subscription not owned by the session itself. And that should be so.
4. I think that unsubscribing from the unexisting subscription should not be a protocol violation even if right now there can be no race condition as I feel that this can lead to possible problems in future extensions. And having ERROR is aligned with our flow in other similar cases.

**Reply by oberstet** [reply]
> In implementations that I'm aware of - it is impossible to unsubscribe from a subscription not owned by the session itself. 

Crossbar.io allows that via meta API - kick a subscriber (on a different session) from a subscription:

https://github.com/crossbario/crossbar/blob/03d7057ee18a0d1e59b33bf1633ee246c1cf908a/crossbar/router/service.py#L549

https://github.com/crossbario/crossbar/blob/03d7057ee18a0d1e59b33bf1633ee246c1cf908a/crossbar/router/service.py#L509

**Reply by KSDaemon** [reply]
Yeah! Meta RPCs are special cases sometimes intended exactly for such things :)

**Comment by KSDaemon** [comment]
@Jopie64 you forgot about wildcards and prefixes. For simple exact topics, your scheme may be ok :) 
But when we deal with pattern-based subscription - things get more complicated. 
Clients subscribe to an URI + matching options. Client A may subscribe to exact URI - this will be subscription 1, Client B may subscribe to prefix-based URI - this will be subscription 2, but they both can receive the same event if it matches URI of first subscription and has the same prefix of the second. 

As an event must be delivered to all clients connected to the router - it is logical that subscriptions are in the router scope, but not in the session. The unsubscribing and unregistering on the other hand always happens in the context of exact session, and router knows what subscriptions that session has.

**Reply by oberstet** [reply]
yes, absolutely. we've arrived at this solution (sub/reg IDs and how they are used) because of this. and also save wire space (transmitting an int in events/calls is better than a whole string identifying a sub/reg)
